### PR TITLE
[android]: fix backlight toggle

### PR DESCRIPTION
**Pull request by zwim**
@pazos: This is a fix for #6515, which works on the Tolinos and on generic Android

But there ~~are two~~ is on problem~~s~~ left:

**Problem 1**
The original Tolino-Reading-App, as well as the top status-line has a toggle button for the backlight. 
It the backlight is **turned OFF** with this button. KOReader cannot change the intensity (neither by gestures nor by the native dialog). Koreader works (f.e. popups showing intensity change), but the intensity is NOT changed, it is always zero.
if the backlight is **tuned ON** with this button, KOReader works as expected.

Turning backlight off with that button logcat shows:
```
D/wingAndyBsystem( 2486): setFrontLightState with: on:false and fromButton:true
D/wingAndyBsystem( 2486): setFrontLightState with: on:false and fromButton:false
D/PhoneStatusBarPolicy( 2536): received intent :android.intent.action.FL_CHANGED
D/PhoneStatusBarPolicy( 2536): update:E80K00
D/PhoneStatusBarPolicy( 2536): received intent :android.intent.action.FL_CHANGED
D/PhoneStatusBarPolicy( 2536): update:E80K00
W/ResourceType( 2536): Requesting resource 0x7f06000a failed because it is complex
```

Turning backlight on with that button, logcat shows:
```
D/wingAndyBsystem( 2486): setFrontLightState with: on:true and fromButton:true
D/wingAndyBsystem( 2486): setFrontLightState with: on:true and fromButton:false
D/PhoneStatusBarPolicy( 2536): received intent :android.intent.action.FL_CHANGED
D/PhoneStatusBarPolicy( 2536): update:E80K00
D/PhoneStatusBarPolicy( 2536): received intent :android.intent.action.FL_CHANGED
D/PhoneStatusBarPolicy( 2536): update:E80K00
W/ResourceType( 2536): Requesting resource 0x7f060009 failed because it is complex
```

~~**Problem 2:**
On my phone. Principally intensity changes work. But if I drag the intensity to zero (native dialog or gesture), then KOReader shows  the right intensity (f.e. popups). But the intensity is the intensity used by the system.~~ This problem is solved with https://github.com/koreader/android-luajit-launcher/pull/246

@pazos: Do you have an idea how to fix both problems?

---
This change is [Reviewable](https://reviewable.io/reviews/koreader/koreader/6520)

### CODE REVIEW

**Comment by zwim** [comment]
@pazos: I have broken down the frontlight problem on Tolinos by looking into the Tolino.apk

First I found in `SettingsIlluminationPresenter.class`:
```
  public void setFrontlightState(boolean paramBoolean) {
    if (!EPubApplication.isGoingToSleep) {
      ReflectionUtil.setFrontlightState((Context)this.activity, paramBoolean);
      ReflectionUtil.handleSunIconInStatusBarVisibility((Context)this.activity);
    } 
  }
``` 

In the `Reflectionutil.class` there is:
```
  public static void setFrontlightState(Context paramContext, boolean paramBoolean) {
    PowerManager powerManager = (PowerManager)paramContext.getSystemService("power");
    try {
      powerManager.getClass().getMethod("setFrontLightState", new Class[] { boolean.class }).invoke(powerManager, new Object[] { Boolean.valueOf(paramBoolean) });
      if (ApplicationEnviromentProperties.isDebugMode())
        Log.d(LOG_TAG, "setFrontlightState: " + paramBoolean); 
      return;
    } catch (IllegalArgumentException illegalArgumentException) {
      illegalArgumentException.printStackTrace();
      return;
    } catch (IllegalAccessException illegalAccessException) {
      illegalAccessException.printStackTrace();
      return;
    } catch (InvocationTargetException invocationTargetException) {
      invocationTargetException.printStackTrace();
      return;
    } catch (SecurityException securityException) {
      securityException.printStackTrace();
      return;
    } catch (NoSuchMethodException noSuchMethodException) {
      try {
        securityException.getClass().getMethod("setFrontLightState", new Class[] { boolean.class, boolean.class }).invoke(securityException, new Object[] { Boolean.valueOf(paramBoolean), Boolean.valueOf(false) });
        if (ApplicationEnviromentProperties.isDebugMode()) {
          Log.d(LOG_TAG, "setFrontlightState: " + paramBoolean);
          return;
        } 
        return;
      } catch (Exception exception) {
        exception.printStackTrace();
        return;
      } 
    } 
  }
```
and
```
public static boolean getFrontlightState(Context paramContext) {
    PowerManager powerManager = (PowerManager)paramContext.getSystemService("power");
    try {
      Object object = powerManager.getClass().getMethod("isFrontLightOn", new Class[0]).invoke(powerManager, new Object[0]);
      if (ApplicationEnviromentProperties.isDebugMode())
        Log.d(LOG_TAG, "isFrontLightOn: " + Boolean.parseBoolean(object.toString())); 
      return Boolean.parseBoolean(object.toString());
    } catch (IllegalArgumentException illegalArgumentException) {
      illegalArgumentException.printStackTrace();
      return false;
    } catch (IllegalAccessException illegalAccessException) {
      illegalAccessException.printStackTrace();
      return false;
    } catch (InvocationTargetException invocationTargetException) {
      invocationTargetException.printStackTrace();
      return false;
    } catch (SecurityException securityException) {
      securityException.printStackTrace();
      return false;
    } catch (NoSuchMethodException noSuchMethodException) {
      try {
        Object object = securityException.getClass().getMethod("isFrontLightOn", new Class[0]).invoke(securityException, new Object[0]);
        if (ApplicationEnviromentProperties.isDebugMode())
          Log.d(LOG_TAG, "getFrontlightState: " + Boolean.parseBoolean(object.toString())); 
        return Boolean.parseBoolean(object.toString());
      } catch (Exception exception) {
        exception.printStackTrace();
        return false;
      } 
    } 
  }
```

I think we would have to add two methods `getFrontlightState` and `setFrontlightState` to `android.lua` and friends to solve the **Problem 1**.
@pazos: As I am not firm with android, would you like to help me with that?

**Comment by pazos** [comment]
@zwim: problem 2 is easily fixable (0 means use system settings, so is just a matter of ignoring that value and make the minimum value 1)

About the first problem I have no idea. I think you might find problems trying to port features that are only meant to system apps (like the one we found using System.settings.putInt). Certainly catching `android.intent.action.FL_CHANGED` could be a good way to toggle fl based on the button input, but I don't know if that is possible, feasible or interesting.

Feel free to open an issue in android-luajit-launcher to keep track of android specific issues (there are billions :))

**Comment by zwim** [comment]
> @zwim: problem 2 is easily fixable (0 means use system settings, so is just a matter of ignoring that value and make the minimum value 1)

OK, that is fixable ;-)
 
> About the first problem I have no idea. I think you might find problems trying to port features that are only meant to system apps (like the one we found using System.settings.putInt). Certainly catching `android.intent.action.FL_CHANGED` could be a good way to toggle fl based on the button input, but I don't know if that is possible, feasible or interesting.

I see two easy solutions. Which would you prefer?
1.) On Tolino we could check: We can set the intensity. If frontlight is toggled Off, we could read back the value. If the set and the read values differ, we would know, that the frontlight is toggled off.  In this case we show a message to the user, that he/she should toggle it on (with the original app, or the sun symbol in the status line). Not nice for the user (we have a toggle in KOReader and a toggleButton on Android).
2.) We use the `/sys/class/backlight/*/color`-file (different path names on Epos2 and Vision4) directly. This works despite the toggleButtons on Tolino. Probably better for the user (KOReader works as expected, no need to leave the app).

**Comment by pazos** [comment]
Lets move discussion about Tolinos to https://github.com/koreader/android-luajit-launcher/issues/245 and feel free to open there other tickets related with android brokenness.

File: reader.lua [hunk 1]
```diff
@@ -206,6 +206,14 @@ if G_reader_settings:isTrue("color_rendering") and not Device:hasColorScreen() t
     })
 end
 
+-- Check and warn if frontlight is disabled by system settings (e.g. Tolinos)
```
[the remark this answers is not in the document]
Comment (pazos) [reply reader.lua:209]: I don't know :p.

[the remark this answers is not in the document]
Comment (zwim) [reply reader.lua:209]: @pazos: So the only thing we can do is, to inform the user, that he/she has to toggle the frontlight manually.
If I understood Frenzie correctly, the Infomessage might stay in `reader.lua` (similar to the `Device:neetsTouchScreenProbe`). What do you think?

[the remark this answers is not in the document]
Comment (Frenzie) [reply reader.lua:209]: Assuming you mean the call, not the infomessage. ;-)

[the remark this answers is not in the document]
Comment (zwim) [reply reader.lua:209]: ???
I don`t understand. I mean we should take the commit as it is (without the l10n/ changes).

[the remark this answers is not in the document]
Comment (pazos) [reply reader.lua:209]: You can wrap whatever into a Device function and call it where you want.

But, anyways, it seems pretty invasive to show a infomessage each time that happens. IMO you can get the same level of awareness by displaying a warning when the user tries to set a new warmth level. 

>  I mean we should take the commit as it is

I didn't review the rest of the commit yet. But aside from putting things in `reader.lua` the rest are more or less expected and can be refactored later.

[the remark this answers is not in the document]
Comment (Frenzie) [reply reader.lua:209]: I agree. I was commenting more broadly that I don't think something like `if Device.powerd.canChangeFrontlight ~= nil and not Device.powerd.canChangeFrontlight() then` is idiomatic, for future reference more than anything.

The InfoMessage itself would get annoying pretty quickly if you don't want to do that.

**Comment by zwim** [comment]
All the changes should only affect Tolinos!
I have dropped the Infomessage in `reader.lua`. If the frontlightSwitch is off on start, it will be switched on. 
If a user doesn't want frontlight on, the frontlight it can be set to zero in KOReader.
If the Tolino is not rooted, the switch cannot be changed and therefore the intensity in KOReader cannot be changed, too.

Maybe we should mention this behavior in the Wiki?

**Comment by Frenzie** [comment]
@pazos' suggestion seems perfectly fine to me? There's a distinct difference between every time on startup and every time you actively try to change the setting. :-)

**Comment by zwim** [comment]
> @pazos' suggestion seems perfectly fine to me? There's a distinct difference between every time on startup and every time you actively try to change the setting. :-)

So you mean I should call the `toggleFrontlightSwitchOn` on every intensity change?

**Comment by Frenzie** [comment]
I would normally expect picking an intensity to be interpreted as turning on the front light and setting it to that intensity, yes.

**Review by pazos — COMMENTED** [review]
Logic looks fine :+1: 

Naming scheme I don't know, it doesn't bothers me :p

File: frontend/device/kobo/powerd.lua [hunk 2]
```diff
@@ -348,6 +348,7 @@ function KoboPowerD:turnOffFrontlightHW()
         UIManager:broadcastEvent(Event:new("FrontlightStateChanged"))
     end
 end
+
```
Comment (pazos) [inline frontend/device/kobo/powerd.lua:351]: Unrelated to this PR

**Review by pazos — APPROVED** [review]
Aside from the unrelated new line, LGTM

You need to commit changes in luajit-launcher. Since everything is already in master you can change to master branch and add the submodule change to this PR.

Something like:

```
cd platform/android/luajit-launcher
git checkout master
git pull upstream master
cd -
git add platform/android/luajit-launcher
```

Followed by git commit and push :).

Thanks for all the testing :+1:

File: frontend/device/kobo/powerd.lua (hunk 2 above)
Comment (NiLuJe) [reply frontend/device/kobo/powerd.lua:351]: (I'm fine with it, FWIW ;)).

**Comment by zwim** [comment]
@pazos: Thank you for your help, hints and patience. :+1:

**Pull request merged**